Activity Log: add persistent notices

eliorivero opened the pull request:
Displays the most urgent tasks that the user needs to take care of.

#### Testing

1. downgrade the version of some plugins
2. boot https://calypso.live/?branch=add/activity-log-tasklist
3. go to Activity Log

To forge the plugin update error, you need to set the permissions of the plugin directory to 555 so it's not writable and update fails.

#### Updates and dismiss of single element
(The success notice fades away on its own after 5 secs I just clicked the x for the demo)

![tasklist](https://user-images.githubusercontent.com/1041600/39223980-2cae1afe-481b-11e8-873b-fd5f8e2b4802.gif)

#### Load plugin update after page load that just become available
(I edited the plugin to downgrade its version at the same time)

![tasklist2](https://user-images.githubusercontent.com/1041600/39224048-7ff6514a-481b-11e8-956f-1f73fd684a81.gif)

<img width="785" alt="captura de pantalla 2018-04-26 a la s 17 54 22" src="https://user-images.githubusercontent.com/1041600/39331449-0183add8-497b-11e8-8278-ca56cf821a1e.png">

#### Tracking

This introduces several tracking:

When the tasklist is displayed

<img width="371" alt="captura de pantalla 2018-04-26 a la s 14 23 48" src="https://user-images.githubusercontent.com/1041600/39323505-2c400cbe-4963-11e8-98d3-1423f652e2b5.png">

When a plugin is dismissed

<img width="380" alt="captura de pantalla 2018-04-26 a la s 14 24 33" src="https://user-images.githubusercontent.com/1041600/39323506-2cd6a700-4963-11e8-9151-e626ba76a0d2.png">

When all plugins are dismissed

<img width="381" alt="captura de pantalla 2018-04-26 a la s 14 24 43" src="https://user-images.githubusercontent.com/1041600/39323508-2d70d7a8-4963-11e8-9ac4-ac18516cbd96.png">

When a plugin is updated, from a task

<img width="380" alt="captura de pantalla 2018-04-26 a la s 14 42 01" src="https://user-images.githubusercontent.com/1041600/39323509-2df07602-4963-11e8-8a8a-3e0db4720a21.png">

and from a notice

<img width="379" alt="captura de pantalla 2018-04-26 a la s 14 43 20" src="https://user-images.githubusercontent.com/1041600/39323510-2e5cfe94-4963-11e8-87aa-fbefe08859d1.png">

### Update Apr 26
When one plugin update starts, all the update buttons and the dismiss all menu are disabled

### Update Apr 27
New link Manage plugins in main ellipsis menu takes users to the plugin management screen in Calypso.
Plugin name is now linked to its section inside plugin management.
Two new event tracking:
- calypso_activitylog_tasklist_manage_plugins When user clicks the **Manage plugins** link
- calypso_activitylog_tasklist_manage_single_plugin When user clicks on the plugin name to go to its page

<img width="570" alt="captura de pantalla 2018-04-27 a la s 15 31 11" src="https://user-images.githubusercontent.com/1041600/39379359-574e3b32-4a31-11e8-9e07-0ca0b999d36a.png">

### Update May 4
New button Update all updates all plugins.
Plugin updates are now fired sequentially following a queue.
New tracking `calypso_activitylog_tasklist_update_plugin_all` when user clicks the button to update all plugins
~~New~~ (refactored from an existing one) tracking `calypso_activitylog_tasklist_update_plugin_from_error` tracks when user clicks the Try again button in the error notice. Previously this was tracked in the `calypso_activitylog_tasklist_update_plugin` passing a property location, but doing it this way is easier to get reports.
Tracking is now no longer coupled with the dispatched action `updatePlugin`, since that's now driven by the queue processing rather than user interaction. Instead, clicks are now tracked directly in the onClick method that the buttons call.

dmsnell commented on client/my-sites/stats/activity-log-tasklist/plugins-to-update.jsx:
```diff
@@ -0,0 +1,56 @@
+		};
+
+		componentWillReceiveProps( nextProps ) {
+			if ( this.state.pluginsWithUpdates.length < nextProps.pluginsWithUpdates.length ) {
```
watch out! this logic is testing an indirect measure and will fail when we get a _different_ set of the _same length_ as the previous set of plugins

eliorivero replied:
Good catch. Thanks. WIll fix.

dmsnell commented on client/my-sites/stats/activity-log-tasklist/plugins-to-update.jsx:
```diff
@@ -0,0 +1,56 @@
+					),
+				} );
+			}
+		}
```
note that this won't run on mount. it would be good to use the new `getDerivedStateFromProps()` function here so that it _will_ and also because `componentWillReceiveProps()` is now officially being deprecated.

```js
getDerivedStateFromProps( nextProps, prevState ) {
	return {
		pluginsWithUpdates: nextProps.pluginsWithUpdates,
		allPluginsToUpdate: unionBy( nextProps.pluginsWithUpdates, prevState.allPluginsToUpdate, 'slug' )
	};
}
```

I'd also like to encourage you to take advantage of the fact that we have multiple levels of namespacing going on here already and that long names aren't always helpful. terse names can make it easy for many people to quickly ascertain the purpose a variable or function serves. in this case, we already know we're dealing with plugins to update because that's the file name and that's also the component name. Something terser can be easier to digest…

```js
getDerivedStateFromProps( nextProps, prevState ) {
	return {
		pending: nextProps.plugins,
		plugins: unionBy( nextProps.plugins, prevState.plugins, 'slug' )
	};
}

return <Wrapped siteId={ siteId } updatablePlugins={ this.state.pending } plugins={ this.state.plugins } />
```

finally, I'm confused on the name `allPluginsToUpdate` because our comments indicate that it involves not only plugins to update but also plugins which _were_ updated and no longer have any updates available. any way we could refine that name to be more specific or more general so that the contradiction disappears?

eliorivero replied:
Cool. I actually tried using `getDerivedStateFromProps` a couple of weeks ago and it wasn't available. I see that react and react-dom are now at 16.3.1 so will update.
Will review the names and about `allPluginsToUpdate`, I have nothing. `allPlugins` is not good because they're not all plugins, just those that will or have been updated now. Full name would be `allPluginsUpdatedOrToUpdate` and it's quite long. I'm open to suggestions since I also think it's not a good enough name.

dmsnell replied:
Remember that we have the file and component as a namespace, so we don't need to be super redundant inside with our names. It's okay to say that something like `allPlugins` or even `plugins` by itself has a specific meaning inside our component/domain.

`allPluginsUpdatedOrToUpdate` is long and redundant and can distract our train of thought when reading code: this why being so precise is not always helpful.

eliorivero replied:
This is now fixed and I've refactored things to get rid of the prop for pending plugin updates. Now there's a single one `plugins` that has the complete list of plugins with pending updates and those that were updated now.

dmsnell commented on client/my-sites/stats/activity-log-tasklist/plugins-to-update.jsx:
```diff
@@ -0,0 +1,56 @@
+
+		render() {
+			return (
+				<WrappedComponent
```
Note here that we're wiping out any props passed to the wrapper and only sending these three. We may want to send `this.props` as well, or those which this wrapper doesn't intercept

```js
const { pluginsWithUpdates, siteId, ...props } = this.props;

return <WrappedComponent { ...props } siteId={ siteId } … />
```

eliorivero replied:
I intentionally wanted to pass only these since I don't need anything else.

dmsnell replied:
It's probably worth a big comment expressing that then or worth reconsidering because I don't think that's what most people will expect when they wrap a component.

I know that the design of this grew out of a single component into a split. If we don't want the wrapper to be used elsewhere we can also consider defining it in the same file as the wrapped component. That still has merit and removes some of the burden of not coupling the two.

eliorivero replied:
Passed all props as `{ ...this.props }` to comply with expectations.

dmsnell commented on client/my-sites/stats/activity-log-tasklist/index.jsx:
```diff
@@ -0,0 +1,290 @@
+ * Internal dependencies
+ */
+import ActivityLogTaskUpdate from './update';
+import alPluginsToUpdate from './plugins-to-update';
```
"activity log" is already in the file name so it should be clear here without the awkward `al` prefix. Also wrappers are commonly named by what they provide "with"

```js
import WithUpdatablePlugins from './plugins-to-update';

export WithUpdatablePlugins( … );
```

eliorivero replied:
👍

dmsnell commented on client/my-sites/stats/activity-log-tasklist/index.jsx:
```diff
@@ -0,0 +1,290 @@
+	static propTypes = {
+		siteId: PropTypes.number,
+		pluginsWithUpdates: PropTypes.arrayOf( PropTypes.object ), // Plugins with pending updates
+		allPluginsToUpdate: PropTypes.arrayOf( PropTypes.object ), // Don't use. Use connected pluginsToUpdate instead.
```
this is a bit concerning. if we aren't supposed to use it, why are we specifying it?

eliorivero replied:
This is a prop passed by the wrapper. I wanted to declare it to have the type checking. This is only used in the `connect`. I'll update the comment since anybody can actually use this if they want to.

dmsnell replied:
If it's for `connect` or the wrapper then we should remove it from the wrapped. That's introducing backwards coupling that will confuse people.

We can either ignore it here or explicitly lift it out of the passed attributes from `connect()`.

If you want the type-checking inside of `connect()` then we can expand the connected component into its own component and add explicit `propTypes` to it.

```js
const innerComponent = connect( … )( Component );
innerComponent.propTypes = { … }

export default innerComponent;
```

dmsnell commented on client/my-sites/stats/activity-log-tasklist/index.jsx:
```diff
@@ -0,0 +1,290 @@
+	dismissAll = () => {
+		// ToDo: this should update some record in the tasklist API
+		this.setState( { dismissAll: true } );
+	};
```
note that if we get new plugins after clicking dismiss all they won't appear.
if we want to though we can track which ones were being dismissed…

```js
dismissAll = () => {
	this.setState( {
		// assuming `pluginsToUpdate` is an object whose keys are slugs
		dismissed: union( this.state.dismissed, Object.keys( this.props.pluginsToUpdate ) )
	} );
}
```

dmsnell commented on client/my-sites/stats/activity-log-tasklist/index.jsx:
```diff
@@ -0,0 +1,290 @@
+		each( Object.values( pluginsToUpdate ), plugin => {
+			const pluginSlug = plugin.slug;
+
+			if ( false === get( this.props.pluginsToUpdate, `${ pluginSlug }.updateStatus`, false ) ) {
```
this is a general note on the use of `get()` and it applies to the other cases, but it's quite awkward here to be mixing around property access mechanisms and I'm not sure if I'm missing something or if there is a reason for it, but if we are going to use `get()` then I'd recommend staying all-in on one pattern (some exceptions like `this.props` or `this.state` are normal)…

```js
get( this.props, [ 'pluginsToUpdate', pluginSlug, 'updateStatus' ] )
```

also note that `get()` returns `undefined` if the value doesn't exist, so it _appears_ that the default value of `false` may be unnecessary

eliorivero replied:
Will test the value. The `updateStatus` is `false` before a update begins. When an update is under progress, it's an object.

I'll update the string in `get`. Can you point a reason not to compose a string like I did? Is it a guideline in Calypso? Perhaps this could be enforced in the pre-commit processor.

dmsnell replied:
It's not something likely easy to lint and there's no single reason not to do so - it's more of a general programming principle. In a single expression we have different coding patterns and it's not even adding any value. If it _were_ adding significant value we could say it's fine as an exception but in this case I'd argue that it's just making things needlessly inconsistent.

`get()` is designed to make deeply-nested property access safe, so if we are passing deeply-nested properties to its first argument we're taking away from its sole responsibility and jumping out of that safety. In this specific case we're fine because `this.props` will always exist and when `pluginsToUpdate` _doesn't_ exist it's still okay because `get()` knows how to handle `undefined` as the object.

`get()` is also designed to work with one of two modes: an array of path-traversals; and a dot-delimited string which it will parse and convert into an array of path-traversals. by using templates literals to build that dot-delimited path we're preemptively undoing what `get()` will do again. in other words, we take the end-result that `get()` wants, disguise it, then make `get()` uncover it to get back to where we started.

if we mainly want a statically-known deeply-nested property it's easy to pass a string

```js
get( data, 'items.camping.tents.0' )
```

if we are dynamically generating that path then the array format is more appropriate

```js
get( data, [ 'items', category, subCategory, 0 ] )
```

dmsnell commented on client/my-sites/stats/activity-log-tasklist/index.jsx:
```diff
@@ -0,0 +1,290 @@
+			const updateNoticeId = get(
+				this.state.pluginUpdateNotice,
+				[ pluginSlug, 'notice', 'noticeId' ],
+				null
```
on the React deprecation of `componentWillReceiveProps()` I believe we would want to put things like this into `componentDidUpdate()` instead as `componentWillReceiveProps()` may actually run multiple times before rendering and that could do funky things with the notices here.

eliorivero replied:
👍

dmsnell commented on client/my-sites/stats/activity-log-tasklist/index.jsx:
```diff
@@ -0,0 +1,290 @@
+
+		const { translate, pluginsWithUpdates } = this.props;
+		const dismissedPlugins = this.state.dismissedPlugins;
+		const pluginsToUpdate = Object.values( omit( this.props.pluginsToUpdate, dismissedPlugins ) );
```
this statement may be enough of an indication that we have some work still left to do on defining our data structure. I think we'll discover that as-is it's making it hard for someone to understand what the data should look like and how to use it.

dmsnell commented on client/my-sites/stats/activity-log-tasklist/index.jsx:
```diff
@@ -0,0 +1,290 @@
+		const { translate, pluginsWithUpdates } = this.props;
+		const dismissedPlugins = this.state.dismissedPlugins;
+		const pluginsToUpdate = Object.values( omit( this.props.pluginsToUpdate, dismissedPlugins ) );
+		const numberOfPluginUpdates = pluginsWithUpdates.length - dismissedPlugins.length;
```
also presumably this is just `pluginsToUpdate.length`

eliorivero replied:
Not really. `pluginsToUpdate` includes all plugins, updated or pending. `pluginsWithUpdates` includes those that have pending updates, and `dismissedPlugins` should be subtracted since they might have been dismissed and not updated.

eliorivero commented on client/my-sites/stats/activity-log-tasklist/index.jsx:
```diff
@@ -0,0 +1,290 @@
+		);
+
+		if ( updateNoticeId ) {
+			removeThisNotice( updateNoticeId );
```
Ah fantastic, I should have read the readme better.

eliorivero commented:
@dmsnell I've addressed the issues and also refactored things to remove a lot of redundant code like the list of plugins with only pending updates. The method that dismissed all updates is gone too now replaced by the method that dismissed a single plugin that can now also dismiss all, and following the same pattern of adding the dismissed plugins to an array instead of setting a new state prop like the previous dismissAll.

keoshi commented:
Added a couple of style tweaks on https://github.com/Automattic/wp-calypso/pull/24453/commits/824fbbe8feae36ec9e8118323634e5b88fa070fa — nothing major as you made it look great from the get go, @eliorivero. Thank you so much for your work here!

One thing I noticed is that the plugins list is not clear once an update is successful. Notice `Newsletter` here:

![image](https://user-images.githubusercontent.com/390760/39353955-12802dec-4a01-11e8-9e78-61b34f3ad8e9.png)

The update notice is then cleared if I refresh the page.

Also, should we add an item to the ellipsis menu with “Update all”?

Awesome stuff!! 🙌

keoshi commented:
After more testing I also noticed that after dismissing a plugin update, it comes back when the page is refreshed:

![al-dismiss-plugin-update](https://user-images.githubusercontent.com/390760/39354194-c63c19f4-4a01-11e8-8234-0072c5cb2ab7.gif)

eliorivero commented:
> One thing I noticed is that the plugins list is not clear once an update is successful

There was a regression somewhere. This was working like I show in the first gif: after the update, the row goes away. I'll fix this.

>  after dismissing a plugin update, it comes back when the page is refreshed

Yes, for now, since we need to save it in the API later. Or maybe we can do it now with some persistent state saved in Calypso.

eliorivero commented:
Gret suggestion @jeffgolenski! The *Dismiss all* link was a bit lonely there. Will add the menu item.

beaulebens commented:
Since we're adding things left and right ( :P ), and along the lines of Jeff's suggestion -- it seems like the plugin name should be linked (same as it is in a normal event) over to the plugin details for that specific plugin (where they'd also be able to update it directly).

eliorivero commented:
@jeffgolenski I've introduced to suggestion and also linked the plugin name to its page in Calypso. There's new tracking for both events.

<img width="570" alt="captura de pantalla 2018-04-27 a la s 15 31 11" src="https://user-images.githubusercontent.com/1041600/39379359-574e3b32-4a31-11e8-9e07-0ca0b999d36a.png">

@keoshi I've fixed the issue of the update row not vanishing, it's working like before now. Good catch. I was able to reduce quite a bit of code removing unnecessary code related to this.

The PR description is updated to include the additions and the event tracking names.

a8ci18n commented on client/my-sites/stats/activity-log-tasklist/index.jsx:
```diff
@@ -0,0 +1,370 @@
+							className="activity-log-tasklist__menu-item"
+						>
+							<Gridicon icon="cog" size={ 24 } />
+							<span>{ translate( 'Manage plugins' ) }</span>
```
Hi! I've found a possible matching string that has already been translated **66** times:
`translate( 'Manage Plugins' )` *ES Score: 17*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

dmsnell commented on client/my-sites/stats/activity-log-tasklist/index.jsx:
```diff
@@ -0,0 +1,414 @@
+	dequeuePlugin = () =>
+		this.setState(
+			{
+				queuedPlugins: drop( this.state.queuedPlugins ),
```
it's fine to use but there's no need to pull in `lodash` here. the normative form in straight-up JS is…

```js
this.state.queuedPlugins.slice( 1 )
```

eliorivero replied:
Fantastic 👍

dmsnell commented on client/my-sites/stats/activity-log-tasklist/index.jsx:
```diff
@@ -0,0 +1,414 @@
+			setTimeout(
+				() => page.replace( `/stats/activity/${ this.props.siteSlug }`, null, false, false ),
+				10
+			);
```
what's the point of the timeout here?

eliorivero replied:
No reason experienced. Will remove.

dmsnell commented on client/my-sites/stats/activity-log-tasklist/index.jsx:
```diff
@@ -0,0 +1,414 @@
+		page.exit( '*', ( context, next ) => {
+			if (
+				! this.state.queuedPlugins.length ||
+				window.confirm( this.props.translate( 'Cancel plugin updates and leave?' ) )
```
We can't cancel plugin updates can we?

eliorivero replied:
Not the one currently in progress, but others in the queue will be canceled.

eliorivero replied:
Well, the customer expects all plugins to be updated, which won't happen if they navigate away.

eliorivero commented on client/my-sites/stats/activity-log-tasklist/index.jsx:
```diff
@@ -0,0 +1,414 @@
+	};
+
+	componentDidMount() {
+		page.exit( '*', ( context, next ) => {
```
This was taken from 
https://github.com/Automattic/wp-calypso/blob/4b11becda3fc7dec39f5a87832bb2175ebd1e3c2/client/my-sites/plugins/jetpack-plugins-setup/index.jsx#L106
Preventing navigating away during a batch update is something we agreed with @samhotchkiss.
Before going with this one, I tried a more strict path:
```
`/stats/activity/${ this.props.siteSlug }`
```
and it doesn't make a difference. There is a caveat with all these approaches (and I believe the place where I took this from—which also unnecessarily adds window handlers—also suffers from it) and is that repeated clicks in other tabs or menu items eventually allows the user to navigate without confirmation. It's something I'd like to talk to you about today since I'm not sure if it's a shortcoming of page.js or what.

rodrigoi replied:
having the components know about `page` is a bit problematic. Components shouldn't know about `page`, neither about route changes. It makes everything from testing to reusing the component really difficult. And breaks the pattern of calypso's controllers managing `page`.

But rather than having a function passed down from the controller all the way down to this component, I think that the solution is not doing batch updates on the client. It adds a point of failure that is completely out of our control.

So, I agree with @dmsnell. If we don't have the API support to make batch updates on the server and notify the user of progress (and protect the process from failures) it's better not to have the feature at all. Nagging the user is never a good idea.

eliorivero replied:
The "Update all" really adds value to users since they can click once to update all plugins.

rodrigoi replied:
This is not a critic on the value of the feature. But having this batch operation controlled by the client makes it extremely fragile, with a high risk of failures we can't control.

eliorivero replied:
Yup, sorry, my comment was directed to Dennis' comment above. I would prefer not to do this queue and the sequential update, but there are some limitations in the WP core updater that have been solved in the past with the sequential update. More convo at p1525452144000006-activity-log-slack

jsnajdr replied:
It would be nice if we could use `lib/protect-form` here. It does exactly the same thing and the implementation is almost identical, too. The only downside is that the confirm message is hardcoded ("You have unsaved changes") and is not customizable.

If only we could extend `protectForm` with a customizable message...

eliorivero commented on client/my-sites/stats/activity-log-tasklist/index.jsx:
```diff
@@ -0,0 +1,414 @@
+	keyBy(
+		pluginList.map( plugin =>
+			merge( {}, plugin, { updateStatus: getStatusForPlugin( state, siteId, plugin.id ) } )
+		),
```
Awesome 👍

samhotchkiss commented:
Tests well for me.  I think we are good to ship this, and we can start tracking usage, and continue to iterate and improve.

jsnajdr commented on client/my-sites/stats/activity-log-tasklist/index.jsx:
```diff
@@ -0,0 +1,413 @@
+			) {
+				return next();
+			}
+			page.replace( path, null, false, false );
```
The `setTimeout( () => page.replace( ... ), 0 )` wrapper is most likely needed. All other Calypso implementations of the "navigation blocker" use it, and it's used also in internal implementation of `page.redirect`

eliorivero replied:
Ok I've restored it. Yes, the place I referenced this from in Calypso, had that setTimeout to 0. I tried with and without it, and experienced no difference, but I'm happy to leave it since it's in the other implementations.

a8ci18n commented on client/my-sites/stats/activity-log-tasklist/index.jsx:
```diff
@@ -0,0 +1,420 @@
+							className="activity-log-tasklist__menu-item"
+							icon="cog"
+						>
+							<span>{ translate( 'Manage plugins' ) }</span>
```
Hi! I've found a possible matching string that has already been translated **66** times:
`translate( 'Manage Plugins' )` *ES Score: 17*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).
